siwe.go

```diff
@@ -0,0 +1,462 @@
+	match := _SIWE_MESSAGE.FindStringSubmatch(message)
+
+	if match == nil {
+		return nil, &ParsingFailed{"Message could not be parsed"}
+	}
+
+	result := make(map[string]interface{})
+	for i, name := range _SIWE_MESSAGE.SubexpNames() {
+		if i != 0 && name != "" && match[i] != "" {
+			result[name] = match[i]
+		}
+	}
+
+	originalAddress := result["address"].(string)
+	parsedAddress := common.HexToAddress(originalAddress)
+	if originalAddress != parsedAddress.String() {
+		return nil, &ParsingFailed{"Address must be in EIP-55 format"}
+	}
+
+	if val, ok := result["resources"]; ok {
+		result["resources"] = strings.Split(val.(string), "\n- ")[1:]
+	}
+
+	return result, nil
+}
+
+func ParseMessage(message string) (*Message, error) {
+	result, err := parseMessage(message)
+	if err != nil {
+		return nil, err
+	}
+
+	parsed, err := InitMessage(
+		result["domain"].(string),
+		result["address"].(string),
+		result["uri"].(string),
+		result["version"].(string),
+		result,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return parsed, nil
+}
+
+func signHash(data []byte) common.Hash {
+	msg := fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", len(data), data)
+	return crypto.Keccak256Hash([]byte(msg))
+}
+
+func (m *Message) ValidNow() (bool, error) {
+	return m.ValidAt(time.Now().UTC())
+}
+
+func (m *Message) ValidAt(when time.Time) (bool, error) {
+	if m.expirationTime != nil {
+		if time.Now().UTC().After(*m.getExpirationTime()) {
+			return false, &ExpiredMessage{"Message expired"}
+		}
+	}
+
+	if m.notBefore != nil {
+		if time.Now().UTC().Before(*m.getNotBefore()) {
+			return false, &InvalidMessage{"Message not yet valid"}
+		}
+	}
+
+	return true, nil
+}
+
+func (m *Message) VerifyEIP191(signature string) (*ecdsa.PublicKey, error) {
+	if isEmpty(&signature) {
+		return nil, &InvalidSignature{"Signature cannot be empty"}
+	}
+
+	// Ref: https://stackoverflow.com/questions/49085737/geth-ecrecover-invalid-signature-recovery-id
+	data := m.PrepareMessage()
+	hash := signHash([]byte(data))
+
+	sigBytes, err := hexutil.Decode(signature)
+	if err != nil {
+		return nil, &InvalidSignature{"Failed to decode signature"}
+	}
+
+	// Ref:https://github.com/ethereum/go-ethereum/blob/55599ee95d4151a2502465e0afc7c47bd1acba77/internal/ethapi/api.go#L442
+	if sigBytes[64] != 27 && sigBytes[64] != 28 {
+		return nil, &InvalidSignature{"Invalid signature bytes"}
+	}
+	sigBytes[64] -= 27
+
+	pkey, err := crypto.SigToPub(hash.Bytes(), sigBytes)
+	if err != nil {
+		return nil, &InvalidSignature{"Failed to recover public key from signature"}
+	}
+
+	address := crypto.PubkeyToAddress(*pkey)
+
+	if address != m.address {
+		return nil, &InvalidSignature{"Signer address must match message address"}
+	}
+
+	return pkey, nil
+}
+
+func (m *Message) Verify(signature string) (*ecdsa.PublicKey, error) {
+	_, err := m.ValidNow()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return m.VerifyEIP191(signature)
+}
+
+func (m *Message) PrepareMessage() string {
```

Comment:
I would make it the built in `String` method

Reply:
Should we keep PrepareMessage as well or prefer each language's toString equivalent exclusively?

Reply:
I think `String` can replace the current `PrepareMessage` because it only serialises the message. It's a bit tricky when you want to provide an interface to generate messages to be signed but we don't have to care about that for now.